File: packages/apple/Sources/BuiltIn/ChatChain.swift
import Foundation
import NaturalLanguage

/// Built-in chain for conversational AI with memory support.
///
/// ```swift
/// let chain = ChatChain(memory: BufferMemory())
/// let result = try await chain.run("Hello!")
/// ```
@available(iOS 15.0, macOS 14.0, *)
public struct ChatChain: Chain {
    public let name = "ChatChain"

    private let model: any LocanaraModel
    private let memory: (any Memory)?
    private let systemPrompt: String

    public init(
        model: (any LocanaraModel)? = nil,
        memory: (any Memory)? = nil,
        systemPrompt: String = "You are a friendly, helpful assistant."
    ) {
        self.model = model ?? LocanaraDefaults.model
        self.memory = memory
        self.systemPrompt = systemPrompt
    }

    /// Detect the dominant language of the input text.
    /// Uses NLLanguageRecognizer only when non-Latin characters are present.
    /// Short Latin-only text defaults to English (avoids misdetection like "hi" → Catalan).
    private func detectLanguage(_ text: String) -> String {
        let hasNonLatin = text.unicodeScalars.contains { scalar in
            let v = scalar.value
            return (v >= 0xAC00 && v <= 0xD7AF) || // Hangul
                   (v >= 0x1100 && v <= 0x11FF) || // Hangul Jamo
                   (v >= 0x3130 && v <= 0x318F) || // Hangul Compat Jamo
                   (v >= 0x3040 && v <= 0x309F) || // Hiragana
                   (v >= 0x30A0 && v <= 0x30FF) || // Katakana
                   (v >= 0x4E00 && v <= 0x9FFF) || // CJK
                   (v >= 0x0600 && v <= 0x06FF) || // Arabic
                   (v >= 0x0400 && v <= 0x04FF) || // Cyrillic
                   (v >= 0x0E00 && v <= 0x0E7F)    // Thai
        }

        guard hasNonLatin else { return "English" }

        let recognizer = NLLanguageRecognizer()
        recognizer.processString(text)
        guard let lang = recognizer.dominantLanguage else { return "English" }
        return Locale(identifier: "en").localizedString(forIdentifier: lang.rawValue) ?? "English"
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        var historyText = ""
        if let memory = memory {
            let entries = await memory.load(for: input)
            historyText = entries.map { "\($0.role.capitalized): \($0.content)" }
                .joined(separator: "\n") + "\n"
        }

        let detectedLang = detectLanguage(input.text)
        let languageInstruction = "IMPORTANT: You MUST reply in \(detectedLang). Do NOT reply in any other language."

        let prompt = try BuiltInPrompts.chat.format([
            "text": input.text,
            "systemPrompt": "System instruction: \(systemPrompt)",
            "history": historyText,
            "languageInstruction": languageInstruction
        ])

        print("[ChatChain] detected language: \(detectedLang)")
        print("[ChatChain] input: \(input.text)")
        let response = try await model.generate(prompt: prompt, config: .conversational)
        let message = response.text.trimmingCharacters(in: .whitespacesAndNewlines)
        print("[ChatChain] output: \(message)")

        let result = ChatResult(
            message: message,
            canContinue: true
        )

        let output = ChainOutput(
            value: result,
            text: message,
            metadata: input.metadata,
            processingTimeMs: response.processingTimeMs
        )

        if let memory = memory {
            await memory.save(input: input, output: output)
        }

        return output
    }

    /// Type-safe execution that returns `ChatResult` directly.
    public func run(_ text: String) async throws -> ChatResult {
        let output = try await invoke(ChainInput(text: text))
        guard let result = output.typed(ChatResult.self) else {
            throw LocanaraError.executionFailed("Unexpected output type from ChatChain")
        }
        return result
    }

    /// Stream a chat response, yielding text chunks as they arrive.
    /// Memory is saved after the stream completes.
    public func streamRun(_ text: String) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            Task { @Sendable in
                do {
                    let input = ChainInput(text: text)
                    var historyText = ""
                    if let memory = memory {
                        let entries = await memory.load(for: input)
                        historyText = entries.map { "\($0.role.capitalized): \($0.content)" }
                            .joined(separator: "\n") + "\n"
                    }

                    let detectedLang = detectLanguage(text)
                    let languageInstruction = "IMPORTANT: You MUST reply in \(detectedLang). Do NOT reply in any other language."

                    let prompt = try BuiltInPrompts.chat.format([
                        "text": text,
                        "systemPrompt": "System instruction: \(systemPrompt)",
                        "history": historyText,
                        "languageInstruction": languageInstruction
                    ])

                    print("[ChatChain] streaming, detected language: \(detectedLang)")
                    print("[ChatChain] input: \(text)")

                    var accumulated = ""
                    for try await chunk in model.stream(prompt: prompt, config: .conversational) {
                        accumulated += chunk
                        continuation.yield(chunk)
                    }

                    print("[ChatChain] output (streamed): \(accumulated)")

                    if let memory = memory {
                        let result = ChatResult(message: accumulated, canContinue: true)
                        let output = ChainOutput(
                            value: result,
                            text: accumulated,
                            metadata: input.metadata,
                            processingTimeMs: nil
                        )
                        await memory.save(input: input, output: output)
                    }

                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }
}

File: packages/android/locanara/src/main/kotlin/com/locanara/builtin/ChatChain.kt
package com.locanara.builtin

import com.locanara.ChatResult
import com.locanara.composable.Chain
import com.locanara.composable.Memory
import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput
import com.locanara.core.GenerationConfig
import com.locanara.core.LocanaraDefaults
import com.locanara.core.LocanaraModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

/**
 * Built-in chain for conversational AI with memory support.
 *
 * ```kotlin
 * val chain = ChatChain(memory = BufferMemory())
 * val result = chain.run("Hello!")
 * println(result.message)
 * ```
 */
class ChatChain(
    private val model: LocanaraModel = LocanaraDefaults.model,
    private val memory: Memory? = null,
    private val systemPrompt: String = "You are a friendly, helpful assistant."
) : Chain {
    override val name = "ChatChain"

    /** Detect the dominant language of the input text using Unicode script analysis. */
    private fun detectLanguage(text: String): String {
        var korean = 0; var japanese = 0; var chinese = 0
        var arabic = 0; var cyrillic = 0; var thai = 0; var latin = 0
        for (ch in text) {
            when {
                ch in '\uAC00'..'\uD7AF' || ch in '\u1100'..'\u11FF' || ch in '\u3130'..'\u318F' -> korean++
                ch in '\u3040'..'\u309F' || ch in '\u30A0'..'\u30FF' -> japanese++
                ch in '\u4E00'..'\u9FFF' -> chinese++
                ch in '\u0600'..'\u06FF' -> arabic++
                ch in '\u0400'..'\u04FF' -> cyrillic++
                ch in '\u0E00'..'\u0E7F' -> thai++
                ch in 'A'..'Z' || ch in 'a'..'z' -> latin++
            }
        }
        val counts = mapOf(
            "Korean" to korean, "Japanese" to japanese, "Chinese" to chinese,
            "Arabic" to arabic, "Russian" to cyrillic, "Thai" to thai, "English" to latin
        )
        return counts.maxByOrNull { it.value }?.takeIf { it.value > 0 }?.key ?: "English"
    }

    override suspend fun invoke(input: ChainInput): ChainOutput {
        var historyText = ""
        memory?.let { mem ->
            val entries = mem.load(input)
            historyText = entries.joinToString("\n") {
                "${it.role.replaceFirstChar { c -> c.uppercase() }}: ${it.content}"
            } + "\n"
        }

        val detectedLang = detectLanguage(input.text)
        val languageInstruction = "IMPORTANT: You MUST reply in $detectedLang. Do NOT reply in any other language."

        val prompt = BuiltInPrompts.chat.format(
            mapOf(
                "text" to input.text,
                "systemPrompt" to "System instruction: $systemPrompt",
                "history" to historyText,
                "languageInstruction" to languageInstruction
            )
        )

        println("[ChatChain] detected language: $detectedLang")
        println("[ChatChain] input: ${input.text}")
        val response = model.generate(prompt, GenerationConfig.CONVERSATIONAL)
        val message = response.text.trim()
        println("[ChatChain] output: $message")

        val result = ChatResult(
            message = message,
            canContinue = true
        )

        val output = ChainOutput(
            value = result,
            text = message,
            metadata = input.metadata,
            processingTimeMs = response.processingTimeMs
        )

        memory?.save(input, output)

        return output
    }

    /** Type-safe execution that returns [ChatResult] directly. */
    suspend fun run(text: String): ChatResult {
        val output = invoke(ChainInput(text = text))
        return output.typed<ChatResult>()
            ?: throw IllegalStateException("Unexpected output type from ChatChain")
    }

    /**
     * Stream a chat response, yielding text chunks as they arrive.
     * Memory is saved after the stream completes.
     */
    fun streamRun(text: String): Flow<String> = flow {
        val input = ChainInput(text = text)
        var historyText = ""
        memory?.let { mem ->
            val entries = mem.load(input)
            historyText = entries.joinToString("\n") {
                "${it.role.replaceFirstChar { c -> c.uppercase() }}: ${it.content}"
            } + "\n"
        }

        val detectedLang = detectLanguage(text)
        val languageInstruction = "IMPORTANT: You MUST reply in $detectedLang. Do NOT reply in any other language."

        val prompt = BuiltInPrompts.chat.format(
            mapOf(
                "text" to text,
                "systemPrompt" to "System instruction: $systemPrompt",
                "history" to historyText,
                "languageInstruction" to languageInstruction
            )
        )

        println("[ChatChain] streaming, detected language: $detectedLang")
        println("[ChatChain] input: $text")

        var accumulated = ""
        model.stream(prompt, GenerationConfig.CONVERSATIONAL).collect { chunk ->
            accumulated += chunk
            emit(chunk)
        }

        println("[ChatChain] output (streamed): $accumulated")

        memory?.let { mem ->
            val result = ChatResult(message = accumulated, canContinue = true)
            val output = ChainOutput(
                value = result,
                text = accumulated,
                metadata = input.metadata,
                processingTimeMs = null
            )
            mem.save(input, output)
        }
    }
}
